ykrt/src/deopt.rs

```diff
@@ -338,8 +338,15 @@ unsafe extern "C" fn __ykrt_deopt(
             // Execute the side trace.
             let f = mem::transmute::<
                 _,
-                unsafe extern "C" fn(*mut c_void, *const CompiledTrace, *const c_void) -> (),
+                unsafe extern "C" fn(*mut c_void, *const CompiledTrace, *const c_void) -> !,
             >(st.entry());
+
+            // This `CompiledTrace` was cloned before passing it into this trace. Since the
+            // function `f` never returns, we need to make sure to drop `ct` here to avoid leaks.
+            // FIXME: If this is the last `Arc`, the `CompiledTrace` is freed at this point, and
+            // `st` will no longer be available when we run the side-trace below.
+            drop(ctr);
```

Comment:
I'm confident this is safe in the sense of just `ctr`, because Rust's type system has proven it safe (plus I checked manually, but i trust that less!).

However I think the `FIXME` is wrong: the `st` is a `Clone` of an `Arc` and stands alone. In other words, I don't think the `ctr` being dropped causes the `st` to be dropped.

Reply:
Oh, right. Because `st` is an `Arc` it can't be dropped, which keeps the `CompiledTrace` alive, even if all it's `Arc`s have been dropped. So I assume as soon as `st` drops, the `CompiledTrace` drops too, but I'm not entirely sure how Rust knows when that happens.

Reply:
Rust doesn't know because of the `Arc::into_raw(st)` -- it's our job to turn it back into a non-raw `Arc` so that it can be dropped appropriately. I don't know if we do that anywhere though?

Reply:
Ah wait now I get it. Since `st` is an `Arc` it can exist even if `CompiledTrace` is freed!

Reply:
I will remove the `FIXME`.

Reply:
Note that `st` is also `Arc<CompiledTrace>` and that we then `Arc::into_raw` it. So it still needs to be `Arc::from_raw`ed at some point.

Reply:
> Note that `st` is also `Arc<CompiledTrace>` and that we then `Arc::into_raw` it. So it still needs to be `Arc::from_raw`ed at some point.

It will be dropped during deopt either by the `from_raw` at the beginning of `__ykrt_deopt` or by the new `drop` added in this PR, since the only way to leave a (side)trace is via deopt.